File: java/stayhome/Stayhome.java
import java.io.*;
import java.util.*;

public class Stayhome {
    public static void main(String args[]) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader(args[0]));
        String line;

        int x = 0, y = 0;

        while ((line = br.readLine()) != null) {
            String[] values = line.split("(?!^)");
            for (y = 0; y < values.length; y++) {
                char c = values[y].charAt(0);
                grid[x][y] = new Node(c, 'A', 1000000);
            }
            x++;
        }

        br.close();

        int N = x, M = y, sta, rt, ho, me, wu, han;
        sta = rt = ho = me = wu = han = -1;

        for (int i = 0; i < N; i++)
            for (int j = 0; j < M; j++)
            {
                int symbol = grid[i][j].getI();
                if (symbol == 'S')
                {
                    sta = i;
                    rt = j;
                }
                if (symbol == 'W')
                {
                    wu = i;
                    han = j;
                }
                if (symbol == 'T')
                {
                    ho = i;
                    me = j;
                }
                if (symbol == 'A')
                {
                    airports.add(new Node(i,j,-1));
                }
            }

        //System.out.println(sta + " " + rt + " " + ho + " " + me + " " + wu + " " + han);
        //System.out.println((char) grid[sta][rt].getI());
        safe = false;
        queue1.add(new Node(wu, han, 2));
        while (!queue1.isEmpty()) {
            Node head = queue1.remove();

            int i = head.getI();
            int j = head.getJ();
            int t = head.getT();

            //System.out.println(i + " " + j);

            if (j > 0)
                expandWuhan(i, j - 1, t);
            if (j < M - 1)
                expandWuhan(i, j + 1, t);
            if (i > 0)
                expandWuhan(i - 1, j, t);
            if (i < N - 1)
                expandWuhan(i + 1, j, t);

        }

        //System.out.println("----------------------------");

        air = true;
        queue2.add(new Node(sta, rt, 1));
        int time = 1;
        while (!queue2.isEmpty()) {
            Node head = queue2.remove();

            int i = head.getI();
            int j = head.getJ();
            int t = head.getT();

            //System.out.println(i + " " + j);

            if (safe && t > time)
                break;

            if (i < N - 1)
                expandSotiris(i + 1, j, 'D', t);
            if (j > 0)
                expandSotiris(i, j - 1, 'L', t);
            if (j < M - 1)
                expandSotiris(i, j + 1, 'R', t);
            if (i > 0)
                expandSotiris(i - 1, j, 'U', t);
            
            time = t;
        }

        if (safe)
        {
            String outstr = "";
            int i = ho, j = me;
            while (true)
            {
                int dire = grid[i][j].getJ();
                if(i == sta && j == rt)
                    break;
                outstr = (char) dire + outstr;
                if (dire == 'D')
                {
                    i = i - 1;
                    j = j;
                }
                else if (dire == 'L')
                {
                    i = i;
                    j = j + 1;
                }
                else if (dire == 'R')
                {
                    i = i;
                    j = j - 1;
                }
                else if (dire == 'U')
                {
                    i = i + 1;
                    j = j;
                }
            }
            System.out.println(time);
            System.out.println(outstr);
        }
        else 
            System.out.println("IMPOSSIBLE");
    }

    static Node[][] grid = new Node[1000][1000];
    static Queue<Node> queue1 = new LinkedList<>();
    static Queue<Node> queue2 = new LinkedList<>();
    static Queue<Node> airports = new LinkedList<>();
    static boolean safe = false;
    static boolean air = true;

    static void expandWuhan (int i, int j, int t)
    {
        int cha = grid[i][j].getI();
        int dire = grid[i][j].getJ();
        int time = grid[i][j].getT();
        if ((cha == '.' || cha == 'S' || cha == 'T' || cha == 'W') && time > t)
        {
            grid[i][j] = new Node (cha, dire, t);
            queue1.add(new Node(i, j, t + 2));
        }else if (cha == 'A' && air)
        {
            while(!airports.isEmpty())
            {
                Node head = airports.remove();
                int x = head.getI();
                int y = head.getJ();
                int temp0 = grid[x][y].getI();
                int temp1 = grid[x][y].getJ();
                int temp2 = grid[x][y].getT();
                if (temp2 > t + 5)
                    grid[x][y] = new Node(temp0, temp1, t + 5);
                queue1.add(new Node(x, y, t + 7));
            }
            air = false;
        }
    }

    static void expandSotiris (int i, int j, int d, int t)
    {
        int cha = grid[i][j].getI();
        int dire = grid[i][j].getJ();
        int time = grid[i][j].getT();
        if (t <= time && cha != 'X')
        {
            if (t == time && dire == 'A' && cha != 'T')
                return;
            if (cha == 'T')
                safe = true;
            else
                grid[i][j] = new Node(cha, dire, t);
            if (dire == 'A')
            {
                queue2.add(new Node(i, j, t + 1));
                grid[i][j] = new Node(cha, d, t);
            }
        }
    }
}

File: python3/stayhome.py
from collections import deque

def solve(f):
    global grid, air, queue, safe
    grid = [list(line) for line in f]

    N, M = len(grid), len(grid[0])

    airports = []

    queue = deque()

    for i in range(N):
        for j in range(M):
            if grid[i][j] == 'S':
                start = i, j
            if grid[i][j] == 'W':
                wuhan = i, j
            if grid[i][j] == 'T':
                home = i, j
            if grid[i][j] == 'A':
                airports.append((i, j))
            grid[i][j] = (grid[i][j], 'A', 1000000)
    
    def expandWuhan (i, j, t):
        global grid, queue, air
        char, dire, time = grid[i][j]
        if (char in ('.', 'S', 'T') and time > t):
            grid[i][j] = (char, dire, t)#dire not W
            queue.append((i, j, t + 2))
        elif (char == 'W' and time > t):
            grid[i][j] = (char, dire, t) #dire not W
            queue.append((i, j, t + 2))
        elif (char == 'A' and air):
            for (x, y) in airports:
                if (grid[x][y][2] > t +5):
                    grid[x][y] = (grid[x][y][0], grid[x][y][1], t + 5)#grid[x][y][1] not W
                queue.append((x, y, t + 7))
            air = False

    def expandSotiris (i, j, d, t):
        global grid, queue, safe
        char, dire, time = grid[i][j]
        if (t <= time and char != 'X'):
            if (t == time and dire == 'A' and char != 'T'):
                return
            if (char == 'T'): safe = True
            else: grid[i][j] = (char, dire, t)
            if (dire == 'A'):
                queue.append((i, j, t + 1))
                grid[i][j] = (char, d, t)

    air = True
    x, y = wuhan
    queue.append((x, y, 2))
    while queue:
        i, j, t = queue.popleft()
        if j > 0:
            expandWuhan(i, j - 1, t)
        if j < M - 1:
            expandWuhan(i, j + 1, t)
        if i > 0:
            expandWuhan(i - 1, j, t)
        if i < N - 1:
            expandWuhan(i + 1, j, t)

    safe = False
    x, y = start
    queue = deque()
    queue.append((x, y, 1))
    time = 1
    while queue:
        i, j, t = queue.popleft()
        if (safe and t > time):
            break
        if i < N - 1:
            expandSotiris(i + 1, j, 'D', t)
        if j > 0:
            expandSotiris(i, j - 1, 'L', t)
        if j < M - 1:
            expandSotiris(i, j + 1, 'R', t)
        if i > 0:
            expandSotiris(i - 1, j, 'U', t)
        time = t

    if safe:
        i, j = home
        x, y = start
        outstr = ''
        while True:
            if (i == x and j == y):
                break
            outstr = grid[i][j][1] + outstr
            if (grid[i][j][1] == 'D'):
                i, j = i - 1, j
            elif (grid[i][j][1] == 'L'):
                i, j = i, j + 1
            elif (grid[i][j][1] == 'R'):
                i, j = i, j - 1
            elif (grid[i][j][1] == 'U'):
                i, j = i + 1, j
            
        print(time)
        print(outstr)
    else:
        print('IMPOSSIBLE')


if __name__ == '__main__':
    import sys
    with open(sys.argv[1], 'rt') as f:
        solve(f)
